pkg/cloud/cloud.go

```diff
@@ -333,17 +337,23 @@ var _ Cloud = &cloud{}
 
 // NewCloud returns a new instance of AWS cloud
 // It panics if session is invalid
-func NewCloud(region string, awsSdkDebugLog bool, userAgentExtra string, batching bool) (Cloud, error) {
-	c := newEC2Cloud(region, awsSdkDebugLog, userAgentExtra, batching)
+func NewCloud(region string, awsSdkDebugLog bool, userAgentExtra string, batching bool, roleArn string) (Cloud, error) {
+	c := newEC2Cloud(region, awsSdkDebugLog, userAgentExtra, batching, roleArn)
 	return c, nil
 }
 
-func newEC2Cloud(region string, awsSdkDebugLog bool, userAgentExtra string, batchingEnabled bool) Cloud {
+func newEC2Cloud(region string, awsSdkDebugLog bool, userAgentExtra string, batchingEnabled bool, roleArn string) Cloud {
 	cfg, err := config.LoadDefaultConfig(context.Background(), config.WithRegion(region))
 	if err != nil {
 		panic(err)
 	}
 
+	if roleArn != "" {
+		cfg.Credentials = stscreds.NewAssumeRoleProvider(sts.NewFromConfig(cfg), roleArn, func(aro *stscreds.AssumeRoleOptions) {
+			aro.Duration = assumeRoleDuration
+		})
```

Comment:
not too sure about directly setting `cfg.Credentials` here, so `cfg` was created using the default config - which corresponds to a role (A) in our service account, this role would allow us to assume the customer provided role (B)

would changing the `cfg` impact how the AssumeRoleProvider tries to assume role again after the session expires in an hour? i.e. would the STS client now also attempt to use credential for role (B) (that has expired) to assume the role (B) again?

Reply:
would it make more sense to do sth similar to ENO here and create an `aws.Config` for customer role? i'm thinking something like this

create ec2Cfg using the role here https://github.com/sarthakkothari/aws-ebs-csi-driver/blob/e6a95cbd67fb7e1612c4a0a4839a526b273cf054/pkg/cloud/cloud.go#L351
```
if roleArn != "" {
  ec2Cfg := // create the aws.Config
} else {
  ec2Cfg := cfg
}
```

and use this when creating ec2API in https://github.com/sarthakkothari/aws-ebs-csi-driver/blob/e6a95cbd67fb7e1612c4a0a4839a526b273cf054/pkg/cloud/cloud.go#L368

```
svc := ec2.NewFromConfig(ec2Cfg, // ...
```

Reply:
I understand what you're saying, my thoughts would be what if there are other things in the `cfg` that are used by the EC2 which we are not aware of?